engine/grid.py

```diff
@@ -300,73 +307,164 @@ def find_border_nodes(self):
             'border_nodes' will be initialized.
         """
         border_list = []
-        cols = self.nodes.shape[1]
-        rows = self.nodes.shape[0]
         leftmost_node = None
+        rightmost_node = None
         leftmost_node_pos = None
-        for row in range(rows):
-            for col in range(cols):
+        rightmost_node_pos = None
+        
+        for row in range(self.num_rows):
+            for col in range(self.num_cols):
                 node = self.nodes[row][col]
-                if node.is_active and self.is_on_border(row, col, rows-1, cols-1):
+                #border node 1 left active node 
+                if node.is_active and self.is_on_border(row, col, self.num_rows-1, self.num_cols-1):
                     # check if this is an active node and on the border
                     self.nodes[row][col].is_border = True
                     border_list.append((node, row, col))
                     if leftmost_node_pos is None or col < leftmost_node_pos[1]:
```

Comment:
when updating leftmost_node_pos and leftmost_node, why are we concerned about the col and not the row?